host/sgx/sgxload.c

```diff
@@ -27,9 +27,68 @@
 #include <openenclave/internal/utils.h>
 #include "../memalign.h"
 #include "../signkey.h"
+#include "cpuid.h"
 #include "enclave.h"
 #include "sgxmeasure.h"
 
+#if defined(__linux__)
+// If cpuid.01h:ECX.XSAVE[bit 26] is 1, the processor supports the XSAVE/XRSTOR
+// processor extended states feature, the XSETBV/XGETBV instructions, and XCR0
+static bool is_xcr0_supported(void)
+{
+    uint32_t eax, ebx, ecx, edx;
+
+    eax = ebx = ecx = edx = 0;
+    oe_get_cpuid(1, 0, &eax, &ebx, &ecx, &edx);
+
+    // Bit 26 of ECX indicates support for XGETBV and XCR0
+    if (!(ecx & (1 << 26)))
+        return false;
+
+    return true;
+}
+
+static uint64_t xgetbv(uint32_t index)
+{
+    uint32_t eax, edx;
+    eax = edx = 0;
+
+    if (is_xcr0_supported())
+    {
+        asm volatile(".byte 0x0f,0x01,0xd0" /* xgetbv */
+                     : "=a"(eax), "=d"(edx)
+                     : "c"(index));
+        return eax + ((uint64_t)edx << 32);
+    }
+    else
+        return 0;
+}
+
+static uint64_t get_valid_xfrm(void)
+{
+    uint32_t eax, ebx, ecx, edx;
+    uint64_t xfrm;
+
+    eax = ebx = ecx = edx = 0;
+    // Get Intel SGX Attributes Enumeration Leaf 0x12, sub-leaf 1 from EDX:ECX
+    oe_get_cpuid(0x12, 1, &eax, &ebx, &ecx, &edx);
+    // Get enclave capability
+    // flags = ((uint64_t)ebx << 32) | eax;
+
+    xfrm = xgetbv(0);
+    if (xfrm)
+    {
+        // If x-feature is supported and enabled by OS, ensure it is also
+        // supported in the enclave.
+        xfrm &= (((uint64_t)edx << 32) | ecx);
```

Comment:
No, we should only support up to the feature set we support, not all xfrm bits returned by `XGETBV`. There is probably some additional checking that we don't do today in terms of sizing the XSAVE area correctly as described in the Intel SDM (IIRC we had an old bug where we arbitrarily bumped up the size of the XSAVE area.)

Reply:
Good info here. My code revised yesterday morning does not use this xfrm. Still having the code for informational purposes to log the x-feature supported by the OS.